### PR TITLE
Update displayed hero path when hiring/dismissing creatures that give move penalty

### LINKED ISSUE #4466: Buying creature that gives terrain penalty from mercenary camp (probably dwellings as well) doesn't invalidate pathfinder cache
As in title - if all your troops are terrain-native and you buy creature giving terrain penalty then pathfinder data does not refresh

I guess witch hut with pathfinding might also be worth checking...

### PR DESCRIPTION
**Pull request by MichalZr6**
~This PR extends this fix https://github.com/vcmi/vcmi/pull/2247/commits/6546242c0360f0e37f6da26355f7bf913f864be2~

Displayed hero path on AdventureMapInterface will now be updated when:
- buying creatures in dwellings (level 2 and above) that give movement penalty
- buying creatures in refugee camps that give movement penalty
- dismissing creatures which should take off movement penalty
- giving artifacts like Angel Wings to Seer's Huts

Edit: Resolves #4466, #3286

### CODE REVIEW

File: client/adventureMap/AdventureMapInterface.cpp [hunk 1]
```diff
@@ -168,7 +166,6 @@ void AdventureMapInterface::show(Canvas & to)
 {
 	CIntObject::show(to);
 	dim(to);
-	LOCPLINT->cingconsole->show(to);
```
Comment (IvanSavenko) [inline client/adventureMap/AdventureMapInterface.cpp:171]: Why was this removed?

File: client/adventureMap/AdventureMapInterface.cpp [hunk 2]
```diff
@@ -159,6 +155,8 @@ void AdventureMapInterface::deactivate()
 
 void AdventureMapInterface::showAll(Canvas & to)
 {
+	refreshDisplayedPath();
+
```
Comment (IvanSavenko) [inline client/adventureMap/AdventureMapInterface.cpp:159]: Might be better to place it in required netpack appliers / CPlayerInterface.
showAll() should be used only to display stuff.

Comment (MichalZr6) [reply client/adventureMap/AdventureMapInterface.cpp:159]: Hi Ivan,

I have some doubts if this is the proper way to deal with that. After digging into the code got some observations:

- Shutting down InfoWindow updates the displayed route properly. That's why it works for level 1 dwellings.

- Interface not updated after closing CWindowObject is getting the update on closing (any) popup window.

It looks like the issue might be related to how the display is refreshed when windows are closed. In the destructor of CWindowObject, I see that show() is called to change the cursor visibility:

```
CWindowObject::~CWindowObject()
{
    if (options & RCLICK_POPUP)
        CCS->curh->show();
}
```
Inside show(), the SDL_ShowCursor() function is called, which seems to be triggering some kind of screen refresh:

```
void CursorHandler::show()
{
    if (showing)
        return;

    showing = true;
    cursor->setVisible(true);
}
```

And:

```
void CursorHardware::setVisible(bool on)
{
    GH.dispatchMainThread([on]()
    {
        if (on)
            SDL_ShowCursor(SDL_ENABLE);
        else
            SDL_ShowCursor(SDL_DISABLE);
    });
}
```

Perhaps there should be some function that calls SDL to be updated just before leaving destructors body ? Or this is due to some threads code ?

I also tried to get it done how `InfoWindow::close()` does it but with no success :(

Let me know which approach we should take to resolve this issue — the current approach or the one related to above.

Comment (dydzio0614) [reply client/adventureMap/AdventureMapInterface.cpp:159]: ok so... it should be fixed like here https://github.com/vcmi/vcmi/pull/3685/commits/96b18b1b8e841bd4f3697e494be68e48a5684ccb

Comment (IvanSavenko) [reply client/adventureMap/AdventureMapInterface.cpp:159]: Why do you need to do this on infowindow closing?

When hero accepts creatures from level 1 dwelling game sends RebalanceStacks pack.
This will in turn call ApplyClientNetPackVisitor::visitRebalanceStacks on client
Which calls CPlayerInterface::garrisonsChanged

Updating path must be done in one of these methods. And `visitRebalanceStacks` does have cl.invalidatePaths()

When hero purchases creatures from a dwelling, game sends InsertNewStack pack, which also results in the same call chain including cl.invalidatePaths()

Judging from your changes, I suspect what needs to be changed is that cl.invalidatePaths should also call that verifyPath method

**Review by dydzio0614 — APPROVED** [review]

**Comment by MichalZr6** [comment]
Perhaps `updatePath` instead `verifyPath` would be more suitable naming ?

File: client/NetPacksClient.cpp [hunk 3]
```diff
@@ -230,30 +231,66 @@ void ApplyClientNetPackVisitor::visitEraseStack(EraseStack & pack)
...
 	cl.invalidatePaths(); //it is possible to remove last non-native unit for current terrain and lose movement penalty
+	
+	if(LOCPLINT)
+	{
+		for(const auto hero : LOCPLINT->localState->getWanderingHeroes())
+		{
+			if(hero->id == pack.army)
+				LOCPLINT->localState->verifyPath(hero);
+		}
+	}
```
Comment (IvanSavenko) [inline client/NetPacksClient.cpp:242]: Perhaps move this copy-pasted code into CClient::invalidatePaths?
